File: src/main/java/mdnet/base/settings/ClientSettings.java
package mdnet.base.settings;

import com.google.gson.annotations.SerializedName;

import java.util.Objects;
import java.util.regex.Pattern;

public final class ClientSettings {
	@SerializedName("max_cache_size_mib")
	private final long maxCacheSizeMib;
	@SerializedName("max_bandwidth_mib_per_hour")
	private final long maxBandwidthMibPerHour;
	@SerializedName("max_burst_rate_kib_per_second")
	private final long maxBurstRateKibPerSecond;
	@SerializedName("client_port")
	private final int clientPort;
	@SerializedName("client_secret")
	private final String clientSecret;
	@SerializedName("threads")
	private final int threads;
	@SerializedName("web_settings")
	private final WebSettings webSettings;

	public ClientSettings() {
		this.maxCacheSizeMib = 20480;
		this.maxBandwidthMibPerHour = 0;
		this.maxBurstRateKibPerSecond = 0;
		this.clientPort = 1200;
		this.clientSecret = "PASTE-YOUR-SECRET-HERE";
		this.threads = 32;
		this.webSettings = new WebSettings();
	}

	public ClientSettings(long maxCacheSizeMib, long maxBandwidthMibPerHour, long maxBurstRateKibPerSecond,
			int clientPort, String clientSecret, int threads, WebSettings webSettings) {
		this.maxCacheSizeMib = maxCacheSizeMib;
		this.maxBandwidthMibPerHour = maxBandwidthMibPerHour;
		this.maxBurstRateKibPerSecond = maxBurstRateKibPerSecond;
		this.clientPort = clientPort;
		this.clientSecret = Objects.requireNonNull(clientSecret);
		this.threads = threads;
		this.webSettings = webSettings;
	}

	public long getMaxCacheSizeMib() {
		return maxCacheSizeMib;
	}

	public long getMaxBandwidthMibPerHour() {
		return maxBandwidthMibPerHour;
	}

	public long getMaxBurstRateKibPerSecond() {
		return maxBurstRateKibPerSecond;
	}

	public int getClientPort() {
		return clientPort;
	}

	public String getClientSecret() {
		return clientSecret;
	}
	public WebSettings getWebSettings() {
		return webSettings;
	}

	public int getThreads() {
		return threads;
	}

	@Override
	public String toString() {
		return "ClientSettings{" + "maxCacheSizeMib=" + maxCacheSizeMib + ", maxBandwidthMibPerHour="
				+ maxBandwidthMibPerHour + ", maxBurstRateKibPerSecond=" + maxBurstRateKibPerSecond + ", clientPort="
				+ clientPort + ", clientSecret='" + "<hidden>" + '\'' + ", threads=" + getThreads() + '}';
	}

	public static boolean isSecretValid(String clientSecret) {
		final int CLIENT_KEY_LENGTH = 52;
		return Pattern.matches("^[a-zA-Z0-9]{" + CLIENT_KEY_LENGTH + "}$", clientSecret);
	}
}

File: src/main/kotlin/mdnet/base/settings/ClientSettings.kt
/*
Mangadex@Home
Copyright (c) 2020, MangaDex Network
This file is part of MangaDex@Home.

MangaDex@Home is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

MangaDex@Home is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this MangaDex@Home.  If not, see <http://www.gnu.org/licenses/>.
 */
package mdnet.base.settings

import com.fasterxml.jackson.databind.PropertyNamingStrategy
import com.fasterxml.jackson.databind.annotation.JsonNaming
import dev.afanasev.sekret.Secret

// client settings are verified correct in Main.kt
@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy::class)
data class ClientSettings(
    val maxCacheSizeInMebibytes: Long = 20480,
    val maxMebibytesPerHour: Long = 0,
    val maxKilobitsPerSecond: Long = 0,
    val clientHostname: String = "0.0.0.0",
    val clientPort: Int = 443,
    val clientExternalPort: Int = 0,
    @field:Secret val clientSecret: String = "PASTE-YOUR-SECRET-HERE",
    val threads: Int = 4,
    val gracefulShutdownWaitSeconds: Int = 60,
    val webSettings: WebSettings? = null,
    val devSettings: DevSettings? = null,
    val experimental: Experimental? = null
)

@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy::class)
data class WebSettings(
    val uiHostname: String = "127.0.0.1",
    val uiPort: Int = 8080
)

@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy::class)
data class DevSettings(
    val isDev: Boolean = false
)

@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy::class)
data class Experimental(
    val maxBufferSizeForCacheHit: Int = 0
)
